typing/typecore.ml

```diff
@@ -2948,59 +2947,107 @@ let rec approx_type env sty =
         let tyl = List.map (approx_type env) ctl in
         newconstr path tyl
       end
-  | Ptyp_poly (_, sty) ->
+  | Ptyp_poly _ ->
       approx_type env sty
   | _ -> newvar ()
 
-let rec type_approx env sexp =
-  match sexp.pexp_desc with
-    Pexp_let (_, _, e) -> type_approx env e
-  | Pexp_fun (p, _, spat, e) ->
-      let marg =
-        if has_local_attr_pat spat then Alloc_mode.local
-        else Alloc_mode.newvar ()
+let type_pattern_approx env spat ty_expected =
+  match spat.ppat_desc with
+  | Ppat_constraint(_, ({ptyp_desc=Ptyp_poly _} as sty)) ->
+      let arg_type_mode =
+        if has_local_attr_pat spat then Alloc_mode.Local
+        else Alloc_mode.Global
       in
-      let mret = Alloc_mode.newvar () in
-      let ty = if is_optional p then type_option (newvar ()) else newvar () in
-      let ret = type_approx env e in
-      newty (Tarrow((p,marg,mret), ty, ret, Cok))
+      let ty_pat =
+        Typetexp.transl_simple_type env false arg_type_mode sty
+      in
+      begin try unify env ty_pat.ctyp_type ty_expected with Unify trace ->
+        raise(Error(spat.ppat_loc, env, Pattern_type_clash(trace, None)))
+      end;
+  | _ -> ()
+
+let rec type_function_approx env loc label spato sexp in_function ty_expected =
+  let has_local, has_poly =
+    match spato with
+    | None -> false, false
```

Comment:
This seems to say that `fun (_ : 'a. 'a -> 'a) -> ...` gets special treatment, while `function (_ : 'a. 'a -> 'a) -> ...` does not. That seems strange.

Reply:
The second syntax isn't supported. Polymorphic annotations are only allowed on `fun` parameters. Allowing them on `function` would probably involve allowing them on `match` in general and doesn't seem that useful.